packages/amino/src/encoding.ts

```diff
@@ -0,0 +1,127 @@
+/** See https://github.com/tendermint/tendermint/commit/38b401657e4ad7a7eeb3c30a3cbf512037df3740 */
+const pubkeyAminoPrefixMultisigThreshold = fromHex("22c1f7e2" /* variable length not included */);
+
+/**
+ * Decodes a pubkey in the Amino binary format to a type/value object.
+ */
+export function decodeAminoPubkey(data: Uint8Array): Pubkey {
+  if (arrayContentStartsWith(data, pubkeyAminoPrefixSecp256k1)) {
+    const rest = data.slice(pubkeyAminoPrefixSecp256k1.length);
+    if (rest.length !== 33) {
+      throw new Error("Invalid rest data length. Expected 33 bytes (compressed secp256k1 pubkey).");
+    }
+    return {
+      type: pubkeyType.secp256k1,
+      value: toBase64(rest),
+    };
+  } else if (arrayContentStartsWith(data, pubkeyAminoPrefixEd25519)) {
+    const rest = data.slice(pubkeyAminoPrefixEd25519.length);
+    if (rest.length !== 32) {
+      throw new Error("Invalid rest data length. Expected 32 bytes (Ed25519 pubkey).");
+    }
+    return {
+      type: pubkeyType.ed25519,
+      value: toBase64(rest),
+    };
+  } else if (arrayContentStartsWith(data, pubkeyAminoPrefixSr25519)) {
+    const rest = data.slice(pubkeyAminoPrefixSr25519.length);
+    if (rest.length !== 32) {
+      throw new Error("Invalid rest data length. Expected 32 bytes (Sr25519 pubkey).");
+    }
+    return {
+      type: pubkeyType.sr25519,
+      value: toBase64(rest),
+    };
+  } else {
+    throw new Error("Unsupported public key type. Amino data starts with: " + toHex(data.slice(0, 5)));
+  }
+}
+
+/**
+ * Decodes a bech32 pubkey to Amino binary, which is then decoded to a type/value object.
+ * The bech32 prefix is ignored and discareded.
+ *
+ * @param bechEncoded the bech32 encoded pubkey
+ */
+export function decodeBech32Pubkey(bechEncoded: string): Pubkey {
+  const { data } = Bech32.decode(bechEncoded);
+  return decodeAminoPubkey(data);
+}
+
+/**
+ * Uvarint encoder for Amino. This is the same encoding as `binary.PutUvarint` from the Go
+ * standard library.
+ *
+ * @see https://github.com/tendermint/go-amino/blob/8e779b71f40d175/encoder.go#L77-L85
+ */
+function encodeUvarint(value: number | string): number[] {
+  const checked = Uint53.fromString(value.toString()).toNumber();
+  if (checked > 127) {
+    throw new Error(
+      "Encoding numbers > 127 is not supported here. Please tell those lazy CosmJS maintainers to port the binary.PutUvarint implementation from the Go standard library and write some tests.",
+    );
+  }
+  return [checked];
+}
+
+/**
+ * Encodes a public key to binary Amino.
+ */
+export function encodeAminoPubkey(pubkey: Pubkey): Uint8Array {
+  if (isMultisigThresholdPubkey(pubkey)) {
+    const out = Array.from(pubkeyAminoPrefixMultisigThreshold);
+    out.push(8); // TODO: What is this?
+    out.push(...encodeUvarint(pubkey.value.threshold));
+    for (const pubkeyData of pubkey.value.pubkeys.map((p) => encodeAminoPubkey(p))) {
+      out.push(18); // TODO: What is this?
```

Comment:
Should we use `0x12` style here?

Reply:
Absolutely. Especially since I did not realize that this is the same value is in the signature's hex dump before your comment. Thanks!